packages/plot/src/eval.rs

```diff
@@ -0,0 +1,253 @@
+        }),
+    )
+    .unwrap();
+    ctx.set_function(
+        String::from("sqrt"),
+        Function::new(|arg| match value_to_float(arg) {
+            Some(float) => Ok(Value::Float(float.sqrt())),
+            None => Err(EvalexprError::expected_float(arg.clone())),
+        }),
+    )
+    .unwrap();
+    ctx.set_function(
+        String::from("log"),
+        Function::new(|arg| {
+            let args = arg.as_tuple()?;
+            match (value_to_float(&args[0]), value_to_float(&args[1])) {
+                (Some(a), Some(b)) => Ok(Value::Float(b.log(a))),
+                (_, _) => Err(EvalexprError::expected_float(arg.clone())),
+            }
+        }),
+    )
+    .unwrap();
+    ctx.set_function(
+        String::from("mod"),
+        Function::new(|arg| {
+            let args = arg.as_tuple()?;
+            match (value_to_float(&args[0]), value_to_float(&args[1])) {
+                (Some(a), Some(b)) => Ok(Value::Float(b % a)),
+                (_, _) => Err(EvalexprError::expected_float(arg.clone())),
+            }
+        }),
+    )
+    .unwrap();
+
+    ctx
+}
+
+// Assign values to variables. This function was designed to work with an arbitrary number of
+// variables (to support 3d plotting)
+fn update_function_context(ctx: &mut HashMapContext, names: &[String], values: &[f64]) {
+    let mut names_iter = names.iter();
+    let mut values_iter = values.iter();
+    while let (Some(name), Some(&value)) = (names_iter.next(), values_iter.next()) {
+        // Only values from float ranges will be used here, fine to unwrap
+        ctx.set_value(name.into(), value.into()).unwrap();
+    }
+}
+
+// Approximate a point at the boundary, given two points and the distance to the boundary
+pub fn get_point_at_boundary(outside: (f64, f64), inside: (f64, f64), y_margin: f64) -> (f64, f64) {
+    let (x, y) = outside;
+    let (a, b) = inside;
+    let (dx, dy) = (x - a, y - b);
+    let q = 1.0 - y_margin / dy;
+    (a + dx * q, b + dy * q)
+}
+
+// Used to make divisions by 0 usable for plotting
+pub fn clamp_eval(eval: f64) -> f64 {
+    if eval == f64::INFINITY {
+        f64::MAX
+    } else if eval == f64::NEG_INFINITY {
+        f64::MIN
+    } else {
+        eval
+    }
+}
+
+// This needs to match the functions added in new_function_context
+pub fn get_var_names(function: &str) -> Vec<String> {
+    let replaced = function
+        .replace("sin", "")
+        .replace("cos", "")
+        .replace("tan", "")
+        .replace("sqrt", "")
+        .replace("log", "")
+        .replace("mod", "");
+
+    let mut vars = vec![];
+    for c in replaced.chars() {
+        if c.is_alphabetic() {
+            let s = c.to_string();
+            if !vars.contains(&s) {
+                vars.push(s);
+            }
+        }
+    }
+    vars
+}
+
+// Verify all the functions provided by the user
+pub fn verify_functions(ctx: &mut PlotContext, var_count: usize) -> Result<(), String> {
```

Comment:
Great that you do this check for better error messages! I've not looked close enough to know for sure, but could we not look at the `evalexpr::error::EvalexprError` instead? Like the FunctionIdentifierNotFound variant